.github/workflows/publish-package.yml

```diff
@@ -0,0 +1,44 @@
+name: Publish npc_gzip package
+
+on:
+  release:
+    types: [published]
+
+jobs:
+  
+  publish:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9"]
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      
+      - name: Install Poetry
+        run: |
+          curl -sSL https://install.python-poetry.org | python3 -
+          poetry --version
+
+      - name: Run Tests
+        run: |
+          poetry install
+          poetry run pytest --junit-xml=junit/test-results-${{ matrix.python-version }}.xml
+          
+      - name: Upload pytest test results
+        if: ${{ !cancelled() }}  # Upload results even if tests fail.
+        uses: actions/upload-artifact@v3
+        with:
+          name: pytest-results-${{ matrix.python-version }}
+          path: junit/test-results-${{ matrix.python-version }}.xml
+
+      - name: Poetry Build & Publish
+        run: |
+          poetry build
+          poetry publish
```

Comment:
This will need to access GitHub Actions secrets to work, because `poetry publish` has to authenticate with PyPI. However that is done, this code will probably have to be expanded slightly to facilitate it. In particular, if environment variables are used, the relevant secrets will have to be placed in them using an `env:` key. I don't know if this needs to be taken care of before the PR is merged or not. In particular, if this CI job is only going to be used in the future, and not for initially publishing the package, then there is no need to make any further changes to this file before merging the PR.

Reply:
Yeah so this is a new one for me. Usually I set a pypi api token as an env var. However, when I logged into pypi I saw a new feature for "trusted publishers". I set up this repo as a trusted publisher which I think means this github action will be able to publish this package without storing an api token.

Let's merge with this as is, and if it doesn't work we know how to quickly fix it.

Reply:
Sounds good! Based on [Using trusted publishing with GitHub Actions](https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/#using-trusted-publishing-with-github-actions), I think using trusted publishing in `publish-package.yml` may require either doing the publish step with [the `pypi-publish` action](https://github.com/marketplace/actions/pypi-publish) instead of by running `poetry publish`, or [taking explicit steps](https://github.com/python-poetry/poetry/issues/7940) to provide `poetry` with the OIDC-based token. (Either way, the *build* could definitely still be done with `poetry build`, though.) Either way, as you say, it can be fixed pretty quickly if it doesn't work.

It occurs to me that fixing it may be even easier if [`workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) is added as a second event trigger. Then, if a release is created on GitHub and publishing fails in a way requires the workflow to be modified to fix, then publishing can be manually reattempted from the Actions tab after fixing it, without having to make a second release on GitHub, or delete and remake the tag, etc. (Jobs that have run already can be re-run, but that would use the workflow as it was for the job, rather than the updated workflow. In contrast, triggering the `workflow_dispatch` event from the Actions tab uses the tip of whatever branch it is run for.) *If* you choose to do this, the change I'm suggesting is:

```diff
 on:
   release:
     types: [published]
+  workflow_dispatch:
```